src/net_processing.cpp

```diff
@@ -468,6 +468,44 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
     return &it->second;
 }
 
+/**
+ * Data stucture for an individual peer. This struct is not protected by
+ * cs_main since it does not contain validation-critical data.
+ *
+ * Memory is owned by shared pointers and this object is destructed when
+ * the refcount drops to zero (in FinalizeNode()).
+ *
+ * TODO: move most members from CNodeState to this structure.
+ * TODO: move remaining application-layer data members from CNode to this structure.
+ */
+struct Peer {
+    /** Same id as the CNode object for this peer */
+    NodeId m_id;
+
+    Peer(NodeId id) : m_id(id) {}
+};
+
+using PeerRef = std::shared_ptr<Peer>;
+
+/**
+ * Map of all Peer objects, keyed by peer id. This map is protected
+ * by the global g_peer_mutex. Once a shared pointer reference is
+ * taken, the lock may be released. Individual fields are protected by
+ * their own locks.
+ */
+Mutex g_peer_mutex;
+static std::map<NodeId, PeerRef> g_peer_map GUARDED_BY(g_peer_mutex);
```

Comment:
in commit d60a16b

Is there any reason to make this a global? This map gets populated in `PeerLogicValidation::InitializeNode`, which is called by connman. `PeerLogicValidation` keeps a pointer to this connman, so it would seem natural to me that it also keeps track of the peer map, e.g. in a (private) member.

Reply:
This was mostly just following the convention of net_processing having its data stored in globals. I've had a look at what would be involved in moving `g_peer_map` and `GetPeer()` to be members of `PeerLogicValidation`. The main problem is that in future commits I want to be able to call `GetPeer()` from functions all across net_processing, and many of those aren't member functions of PLV. That means I'd need to either pass a reference to PLV down to those functions, or make them members of PLV.

Doing that (and moving more of the state currently stored as globals into PLV) seems like a reasonable thing to do, but feels a bit separate from the immediate goal of moving data into `Peer` so I'm inclined to leave it for later. What do you think?

Reply:
If I can get buy-in here from everyone else to move a bunch of the static functions from net_processing into PLV, then I'm very happy to move `g_peer_map` into PLV and move those functions as needed in future PRs.

Reply:
+1 on this. No preference whether it happens here or in a follow-up, but I agree PLV is a better home.

Reply:
I started trying to do this, but it's an enormous change. `Misbehaving()` fetches the `PeerRef`, and `Misbehaving()` is called all over net_processing. Let's defer this change to its own PR (probably a series of PRs since it touches almost all of the functions in net_processing.

Reply:
> I wonder if the ~CNetProcessing destructor should be cleaning up this global map, if this isn't moved into PLV? It looks like we don't clean up mapNodeState there, but that seems like an oversight -- unless there's some good reason I'm missing?

All nodes are cleaned up by `CConnman` during the Shutdown sequence (`Shutdown()` in init.cpp calls `CConnman.StopNodes()`, which calls `FinalizeNode()` for each node). When we come to destruct the `PeerLogicValidation` object both maps should already be empty.

I could add a `PeerLogicValidation` dtor which asserts these maps are empty, but that seems a bit orthogonal to this PR (if we introduce a bug that means the nodes are finalized during shutdown, then they'll exist in both the CNodeState map and the Peer map.

Reply:
@MarcoFalke @theuni @sdaftuar - I've made a branch that puts peer_map inside PeerLogicValidation here: https://github.com/jnewbery/bitcoin/tree/2020-08-peer-plv2. It builds on top of (merged) #19704 and (open) #19791. Let me know what you think. If that 19791 gets merged first, I'll switch this PR to use that branch. If not, I'm happy to do the follow-up work to move g_peer_map into PeerLogicValidation later.

Reply:
PR: #19910